mlir/lib/Dialect/MQTOpt/Transforms/QuantumSinkPushPattern.cpp

```diff
@@ -0,0 +1,352 @@
+/*
+ * Copyright (c) 2023 - 2025 Chair for Design Automation, TUM
+ * Copyright (c) 2025 Munich Quantum Software Company GmbH
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ * Licensed under the MIT License
+ */
+
+#include "mlir/Dialect/MQTOpt/IR/MQTOptDialect.h"
+#include "mlir/Dialect/MQTOpt/Transforms/Passes.h"
+
+#include <algorithm>
+#include <cstddef>
+#include <iterator>
+#include <mlir/Dialect/ControlFlow/IR/ControlFlowOps.h>
+#include <mlir/IR/Block.h>
+#include <mlir/IR/MLIRContext.h>
+#include <mlir/IR/Operation.h>
+#include <mlir/IR/PatternMatch.h>
+#include <mlir/IR/ValueRange.h>
+#include <mlir/Support/LLVM.h>
+#include <mlir/Support/LogicalResult.h>
+#include <unordered_set>
+#include <vector>
+
+namespace mqt::ir::opt {
+
+/**
+ * @brief This pattern is responsible to push Unitary operations through branch
+ * operations into new blocks.
+ *
+ * This can be done in some circumstances, if a branch operation uses the result
+ * of a unitary operation as a block argument.
+ */
+struct QuantumSinkPushPattern final
+    : mlir::OpInterfaceRewritePattern<UnitaryInterface> {
+
+  explicit QuantumSinkPushPattern(mlir::MLIRContext* context)
+      : OpInterfaceRewritePattern(context) {}
+
+  /**
+   * @brief Returns the block distance between a block and an operation that
+   * precedes it.
+   *
+   * @param toCheck The block to check.
+   * @param op The operation to find the distance to.
+   * @return The distance between the block and the operation or -1 if the
+   * operation is not in a predecessor block.
+   */
+  int getDepth(mlir::Block& toCheck, const UnitaryInterface& op) const {
+    auto& originalBlock = *op->getBlock();
+    if (&toCheck == &originalBlock) {
```

Comment:
```suggestion
    auto* originalBlock = op->getBlock();
    if (&toCheck == originalBlock) {
```
Just slightly more idiomatic